### PR TITLE
MTSRE-259 | integrate status reporting the hard way

### PR DESCRIPTION
**Pull request by yashvardhan-kukreja**
## Implementation details

`main.go` runs a central heartbeat reporter loop whose purpose is to periodically report the heartbeat which the addon intends to report. This central heartbeat reporter loop is setup via the following way

https://github.com/openshift/reference-addon/blob/ecfb850b2f8c10a603b8b3000aa3a3f0b11ef4f2/cmd/reference-addon-manager/main.go#L98-L107

https://github.com/openshift/reference-addon/blob/ecfb850b2f8c10a603b8b3000aa3a3f0b11ef4f2/internal/addoninstancesdk/sdk.go#L53

> `ReconcilerWithHeartbeat` is just an interface which can be implemented by a Reconciler struct which happens to have some additional methods required for heartbeat reporting to happen. More on `ReconcilerWithHeartbeat` later :)

With every iteration of the heartbeat reporter loop, latest heartbeat is reported concurrently

https://github.com/openshift/reference-addon/blob/ecfb850b2f8c10a603b8b3000aa3a3f0b11ef4f2/internal/addoninstancesdk/sdk.go#L79-L111

> btw, a heartbeat is a metav1.Condition object only.

`ReconcilerWithHeartbeat` is essentially any `Reconciler` struct which implements some additional methods which are required for by the heartbeat reporter.

https://github.com/openshift/reference-addon/blob/ecfb850b2f8c10a603b8b3000aa3a3f0b11ef4f2/internal/addoninstancesdk/sdk.go#L25-L31

**From the perspective of `ReferenceAddonReconciler`**

The `ReferenceAddonReconciler` implements the `ReconcilerWithHeartbeat` by having the following methods:

https://github.com/openshift/reference-addon/blob/ecfb850b2f8c10a603b8b3000aa3a3f0b11ef4f2/internal/controllers/reference_addon_controller.go#L57-L74

And various sections of Reference Addon's codebase can use the `addoninstancesdk.SetAndSendHeartbeat(..., heartbeat)` function to report the heartbeat which is supposed to be reported. For example, if there's a section in the codebase that spots an error which correlates to the addon being unhealthy, then, that section of codebase can use `addoninstancesdk.SetAndSendHeartbeat(r, someUnhealthyHeartbeat)` to do so.

For example:

https://github.com/openshift/reference-addon/blob/ecfb850b2f8c10a603b8b3000aa3a3f0b11ef4f2/internal/controllers/reference_addon_controller.go#L43-L52

### CODE REVIEW

File: cmd/reference-addon-manager/main.go [hunk 1]
```diff
@@ -92,14 +98,31 @@ func main() {
...
-	// if err = (&controllers.AddonReconciler{
-	// 	Client: mgr.GetClient(),
-	// 	Log:    ctrl.Log.WithName("controllers").WithName("Addon"),
-	// 	Scheme: mgr.GetScheme(),
-	// }).SetupWithManager(mgr); err != nil {
-	// 	setupLog.Error(err, "unable to create controller", "controller", "Addon")
-	// 	os.Exit(1)
-	// }
+	// the following sectiom hooks up a heartbeat reporter with the current addon/operator
+	///////////////////////////////////////////////////// -------------------------------- /////////////////////////////////////////////////////
```
Comment (thetechnick) [inline cmd/reference-addon-manager/main.go:102]: I think these section seperators go a little bit overboard :)

File: config/deploy/rbac.yaml [hunk 2]
```diff
@@ -46,7 +46,26 @@ apiVersion: rbac.authorization.k8s.io/v1
...
+  - patch
+- apiGroups:
+  - "addons.managed.openshift.io"
+  resources:
+  - addons
+  resourceNames:
+  - reference-addon ## name of the addon
+  verbs:
+  - get
+  - list
```
Comment (thetechnick) [inline config/deploy/rbac.yaml:68]: I don't think an Addon should have to interact with it's own Addon object and it should even less concern itself with other Addons on the cluster.

File: internal/controllers/reference_addon_controller.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+		Reason:  "ComponentsDown",
+		Message: "Components X, Y and Z are down!",
+	}
+
+	// approximately 5/10 times report a failure condition (probability ~ 50%) and remaining number of times report a successful condition
+	if rand.Intn(10) >= 5 {
+		utils.CommunicateHeartbeat(r.HeartbeatCommunicatorChannel, failureCondition, log)
+	} else {
+		utils.CommunicateHeartbeat(r.HeartbeatCommunicatorChannel, successfulCondition, log)
+	}
```
Comment (thetechnick) [inline internal/controllers/reference_addon_controller.go:46]: I see that this is useful for testing, but our reference implementation should not do things on random :)

File: internal/controllers/reference_addon_controller.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ReferenceAddonReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1.Pod{}). // just a placeholder for now
```
Comment (thetechnick) [inline internal/controllers/reference_addon_controller.go:53]: This will call the `Reconcile` function for every single `Pod` on the cluster and requires RBAC to work with pods.
If we really need something to register here, we should create our own new reference CRD.

**Review by thetechnick — CHANGES_REQUESTED** [review]

File: config/deploy/rbac.yaml (hunk 2 above)
Comment (yashvardhan-kukreja) [reply config/deploy/rbac.yaml:68]: so, the Addon would be making use of AddonInstance client SDK which, behind the scenes, would need to perform a GET to the Addon and parse the target namespace corresponding to it. Rest, the addon (and its developers) need not explicitly interact with its addon object.

File: internal/controllers/reference_addon_controller.go (hunk 3 above)
Comment (yashvardhan-kukreja) [reply internal/controllers/reference_addon_controller.go:46]: true, I just placed it as a placeholder as of now. But I agree with the randomness not being ideal. Let me add more deterministic status reporting :)

File: internal/controllers/reference_addon_controller.go (hunk 4 above)
Comment (yashvardhan-kukreja) [reply internal/controllers/reference_addon_controller.go:53]: Sure, sounds fair! will update :)

File: cmd/reference-addon-manager/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (yashvardhan-kukreja) [reply cmd/reference-addon-manager/main.go:102]: haha, cool, I'll tone them down :)

File: cmd/reference-addon-manager/main.go [hunk 5]
```diff
@@ -92,14 +98,31 @@ func main() {
...
+		setupLog.Error(err, "unable to setup heartbeat reporter")
+		os.Exit(1)
+	}
+	///////////////////////////////////////////////////// -------------------------------- /////////////////////////////////////////////////////
+
+	if err = (&controllers.ReferenceAddonReconciler{
+		Client:                       mgr.GetClient(),
+		Log:                          ctrl.Log.WithName("controllers").WithName("ReferenceAddon"),
+		Scheme:                       mgr.GetScheme(),
+		HeartbeatCommunicatorChannel: heartbeatCommunicatorCh, // linking the heartbeat communicator channel with the reconciler
```
[the remark this answers is not in the document]
Comment (yashvardhan-kukreja) [reply cmd/reference-addon-manager/main.go:121]: Ohh yeah this makes sense. The way I visualise a convenient way for heartbeat reporting is a central loop running concurrently which is reporting heartbeats periodically, and various sections of the addon's codebase are, from time to time, communicating to that central loop which heartbeat to communicate.

I tried to achieve that via channels and had thought about the concerns of channel congestions and blocks but tried to workaround them via timeout-backed channel-sends. But still the other concerns you mentioned are legit.

After thinking a bit about it, I feel that instead of having a channel (`heartbeatCommunicatorCh`) as the point of communication between various sections of the code and the central loop sending heartbeats, we can instead simply have a pointer to a shared variable which encapsulates which heartbeat to send. And the pointer to this shared variable would be an attribute of the ReferenceAddonReconciler struct. This means that across the codebase in various sections, the pointer to that shared variable can be accessed via the ReferenceAddonReconciler object `r`

If this is sounding a bit complicated then, let me come up with a super quick prototype to demostrate this :)

[the remark this answers is not in the document]
Comment (thetechnick) [reply cmd/reference-addon-manager/main.go:121]: +1 for prototype. :)
Not quite sure why you emphasize pointers so much in best case we have interfaces on the consumer side to decouple everything and make unit testing and refactoring easier.

File: config/deploy/rbac.yaml (hunk 2 above)
Comment (thetechnick) [reply config/deploy/rbac.yaml:68]: Because the client is running within the target namespace (As we install the operator into it).
I would get the namespace from the downward API, just to make everything way less complicated:

E.g.:
```yaml
env:
  - name: MY_POD_NAMESPACE
  valueFrom:
    fieldRef:
    fieldPath: metadata.namespace
```

We can also consider using the new config feature to inject the targetNamespace via the Subscription, but this is also making it more complicated again.

File: cmd/reference-addon-manager/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (yashvardhan-kukreja) [reply cmd/reference-addon-manager/main.go:121]: Ohh yes, scrape off that pointer thingy, I was thinking out loud at that moment, my bad :)

File: config/deploy/rbac.yaml (hunk 2 above)
Comment (yashvardhan-kukreja) [reply config/deploy/rbac.yaml:68]: The env-var population via downwards API's solution sounds pretty neat! Doesn't seem to require a massive overhead for the tenants as well, all they gotta do is just throw 
```
- name: ADDON_TARGET_NAMESPACE
  valueFrom:
    fieldRef:
      fieldPath: metadata.namespace
```
under their addon's deployment.spec.template.spec

**Comment by yashvardhan-kukreja** [comment]
Todo: if the current approach of heartbeat reporting (r.latestHeartbeat, r.GetLatestHeartbeat(), r.SetLatestHeartbeat()) gets approved, then we can proceed with adding concurrency safety to `ReferenceAddonReconciler.latestHeartbeat` via mutex

File: cmd/reference-addon-manager/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (yashvardhan-kukreja) [reply cmd/reference-addon-manager/main.go]: Just a quick explanation in the approach I chose.
So, from now, the `ReferenceAddonReconciler` struct will have an attribute called `latestHeartbeat`. This is the so-called shared variable which was talking about.

The central loop (which would report heartbeat) will be defined under main.go and that loop, at every iteration, will access the `r.latestHeartbeat` (actually the getter method of it), and report that condition.

Whereas in various sections of code, whenever something happens and the Addon starts wanting to report a different heartbeat (say, an unhealthy heartbeat), all it has to do is to call `r.SetHeartbeat(...)`, that's it!
This method will update the `r.latestHeartbeat` value and from the next iteration of central loop, it will start reporting this new heartbeat, until changed again.

File: config/deploy/rbac.yaml (hunk 2 above)
Comment (thetechnick) [reply config/deploy/rbac.yaml:68]: We still need to drop RBAC for addons here :)

File: internal/controllers/reference_addon_controller.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+		Reason:  "ImproperNaming",
+		Message: "The addon resources are improperly named",
+	}
+
+	// if the ReferenceAddon object getting reconciled has the name "reference-addon", only then report a successful heartbeat
+	if req.NamespacedName.Name == "reference-addon" {
+		r.SetLatestHeartbeat(successfulCondition)
+	} else {
+		r.SetLatestHeartbeat(failureCondition)
+	}
```
Comment (thetechnick) [inline internal/controllers/reference_addon_controller.go:44]: Can we just report `Success` for now?
The code here would flip/flop the heartbeat condition back and forth when you create two `ReferenceAddon` objects (one called `reference-addon` the other something else...).

File: internal/utils/temporary_utils.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+func SetupHeartbeatReporter(r ReconcilerWithHeartbeat, mgr manager.Manager, addonName string, handleAddonInstanceConfigurationChanges func(addonsv1alpha1.AddonInstanceSpec)) error {
+	defaultHealthyHeartbeatConditionToBeginWith := metav1.Condition{
+		Type:    "addons.managed.openshift.io/Healthy",
+		Status:  "True",
+		Reason:  "AddonHealthy",
+		Message: fmt.Sprintf("Addon '%s' is perfectly healthy", addonName),
+	}
+	// initialized with a healthy heartbeat condition corresponding to the addon
+	r.SetLatestHeartbeat(defaultHealthyHeartbeatConditionToBeginWith)
```
Comment (thetechnick) [inline internal/utils/temporary_utils.go:46]: I don't quite get this.
An Addon needs to decide on it's own whether it's ready or not, ideally by observing the clusters state and reporting based on that information.
Defaulting the state to success make the first reported state completely useless, as it's not reporting the actual state of the addon.

So Addons will get installed, be Healthy once and than fail directly afterwards....

Defaulting to an Unkown or Unsuccessful state in the beginning makes more sense to me.

File: internal/utils/temporary_utils.go [hunk 8]
```diff
@@ -0,0 +1,118 @@
+// TO BE REMOVED AFTER INTEGRATING ADDON INSTANCE CLIENT SDK
```
Comment (thetechnick) [inline internal/utils/temporary_utils.go:1]: This file is in dire need of some unittests :)

File: internal/utils/temporary_utils.go [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+			if err != nil {
+				return fmt.Errorf("failed to get the AddonInstance configuration corresponding to the Addon '%s': %w", addonName, err)
+			}
+			if !reflect.DeepEqual(currentAddonInstanceConfiguration, latestAddonInstanceConfiguration) {
+				currentAddonInstanceConfiguration = latestAddonInstanceConfiguration
+				handleAddonInstanceConfigurationChanges(currentAddonInstanceConfiguration)
+			}
+
+			// waiting for heartbeat update period for executing the next iteration
+			<-time.After(currentAddonInstanceConfiguration.HeartbeatUpdatePeriod.Duration)
```
Comment (thetechnick) [inline internal/utils/temporary_utils.go:72]: We need to change this up a bit, when an Addon is reporting unhealthy or healthy, this status update should go through immediately and not be blocked by the next heartbeat timeout.

The HeartbeatUpdatePeriod is there to update "at least as often", it's not meant as a limit to delay status updates in between.

**Review by thetechnick — CHANGES_REQUESTED** [review]

File: config/deploy/rbac.yaml (hunk 2 above)
Comment (yashvardhan-kukreja) [reply config/deploy/rbac.yaml:68]: Whoops, forgot to remove that, will do :)

File: internal/controllers/reference_addon_controller.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	refapisv1alpha1 "github.com/openshift/reference-addon/apis/reference/v1alpha1"
+)
+
+type ReferenceAddonReconciler struct {
+	client.Client
+	Log             logr.Logger
+	Scheme          *runtime.Scheme
+	latestHeartbeat metav1.Condition
```
Comment (ashishmax31) [inline internal/controllers/reference_addon_controller.go:19]: Don't we want to guard `latestHeartbeat` using a mutex, since the `ReferenceAddonReconciler` object is shared across go-routines (on each reconcile invocation)?

File: internal/controllers/reference_addon_controller.go (hunk 6 above)
Comment (yashvardhan-kukreja) [reply internal/controllers/reference_addon_controller.go:44]: Shouldn't ideally reference-addon conditionally report both success and failure, just so as to demostrate a more realistic scenario?

Although, I agree that we can tune the logic a bit. Maybe we can go ahead with

```
if req.NamespacedName.Name begins with "redhat-*" or "reference-addon-*"
    report Success
else
    report Failure
```

wdyt :)

File: internal/controllers/reference_addon_controller.go (hunk 10 above)
Comment (yashvardhan-kukreja) [reply internal/controllers/reference_addon_controller.go:19]: https://github.com/openshift/reference-addon/pull/5#issuecomment-986844956 :)

File: internal/utils/temporary_utils.go (hunk 7 above)
Comment (yashvardhan-kukreja) [reply internal/utils/temporary_utils.go:46]: Good point!

My perspective was that for any addon in production, it would have passed the osde2e tests in the past because if not, then, it wouldn't have been gated to production in the first place. And, succeeding osde2e tests are indicative of the fact that at least the addon successfully gets installed and bootstrapped on an OSD cluster.

But yeah this poses a risk around the edge case that if an addon with not-so-good heartbeat reporting setup, it might fail, say, just after 30 seconds of getting installed, and say, and still the heartbeat-reporter-central-loop might continue to report the default healthy heartbeat.

Thanks for raising this, I'll make the updates.

That's why it felt like a safe assumption to make that for the first time, it's ok to report a successful heartbeat. But yeah, if the addon starts falling apart, say, 30 seconds after its installation, it will (should) still have the logic to report a failure heartbeat, and if that's not the case, then it will atleast end up reporting HeartbeatTimeout.

File: internal/utils/temporary_utils.go (hunk 8 above)
Comment (yashvardhan-kukreja) [reply internal/utils/temporary_utils.go:1]: Yes! I planned to write the unit tests after the implementation was accepted for this status reporting. Otherwise, after every review, the tests would have required to get refactored as well :)

File: internal/controllers/reference_addon_controller.go (hunk 10 above)
Comment (ashishmax31) [reply internal/controllers/reference_addon_controller.go:19]: Sorry my bad!

File: internal/utils/temporary_utils.go (hunk 9 above)
Comment (yashvardhan-kukreja) [reply internal/utils/temporary_utils.go:72]: I guess a straightforward approach to do this is, modify the `SetLatestHeartbeat(metav1.Condition)` method to
```go
func (r *ReferenceAddonReconciler) SetLatestHeartbeat(heartbeat metav1.Condition) {
	r.latestHeartbeat = heartbeat
        addonName := "reference-addon" // can be also automatically picked by Downwards API from addon's Deployment resource?
        SetAddonInstanceCondition(ctx, r.Client, r,GetLatestHeartbeat(), addonName) // error handling to be added to
}
```

Comment (yashvardhan-kukreja) [reply internal/utils/temporary_utils.go:72]: Although this ends up relying on addon developer's responsibility to do so. Let me think if there's some simple solution to abstract away this real-time heartbeat reporting under the central-heartbeat-reporter-loop itself.

**Comment by yashvardhan-kukreja** [comment]
closing in favor of #8 and #9
